dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java

```diff
@@ -0,0 +1,122 @@
+
+/** Annotates {@link LinkageProblem}s with {@link LinkageProblemCause}s. */
+public class LinkageProblemCauseAnnotator {
+
+  /**
+   * Annotates the cause field of {@link LinkageProblem}s with the {@link LinkageProblemCause}.
+   *
+   * @param rootResult the class path used for generating the linkage problems.
+   * @param linkageProblems linkage problems to annotate
+   * @throws IOException when there is a problem reading JAR files.
+   */
+  static void annotate(ClassPathResult rootResult, Iterable<LinkageProblem> linkageProblems)
+      throws IOException {
+
+    for (LinkageProblem linkageProblem : linkageProblems) {
+      ClassFile sourceClass = linkageProblem.getSourceClass();
+      ClassPathEntry sourceEntry = sourceClass.getClassPathEntry();
+
+      Artifact sourceArtifact = sourceEntry.getArtifact();
+
+      // Resolves the dependency graph with the source artifact at the root.
+      ClassPathBuilder classPathBuilder = new ClassPathBuilder();
+      ClassPathResult subtreeResult = classPathBuilder.resolveWithMaven(sourceArtifact);
+
+      ClassPathEntry entryInSubtree =
+          findClassPathEntryForSymbol(subtreeResult, linkageProblem.getSymbol());
+      if (entryInSubtree == null) {
+        linkageProblem.setCause(UnknownCause.getInstance());
+      } else {
+        Artifact artifactInSubtree = entryInSubtree.getArtifact();
+        DependencyPath pathToSourceEntry = rootResult.getDependencyPaths(sourceEntry).get(0);
+        DependencyPath pathFromSourceEntryToUnselectedEntry =
+            subtreeResult.getDependencyPaths(entryInSubtree).get(0);
+        DependencyPath pathToUnselectedEntry =
+            pathToSourceEntry.concat(pathFromSourceEntryToUnselectedEntry);
+
+        ClassPathEntry selectedEntry =
+            findEntryByArtifactId(
+                rootResult, artifactInSubtree.getGroupId(), artifactInSubtree.getArtifactId());
+        if (selectedEntry != null) {
+          Artifact selectedArtifact = selectedEntry.getArtifact();
+          if (!selectedArtifact.getVersion().equals(artifactInSubtree.getVersion())) {
+            // Different version of that artifact is selected in rootResult
+            linkageProblem.setCause(
+                new DependencyConflict(
+                    rootResult.getDependencyPaths(selectedEntry).get(0), pathToUnselectedEntry));
+          } else {
+            // A linkage error was already there when sourceArtifact was built.
+            linkageProblem.setCause(UnknownCause.getInstance());
+          }
+        } else {
+          // No artifact that matches groupId and artifactId in rootResult.
+          linkageProblem.setCause(new MissingDependency(pathToUnselectedEntry));
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns the class path entry in {@code classPathResult} for the artifact that matches {@code
+   * groupId} and {@code artifactId}. {@code Null} if no matching artifact is found.
+   */
+  private static ClassPathEntry findEntryByArtifactId(
```

Comment:
findEntryById since it's not just the artifact ID

Reply:
Updated.